data/man/subtler.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SUBTLER" "1" "May 2011" "" ""
.
.SH "NAME"
\fBsubtler\fR \- a commandline interface for subtle
.
.SH "SYNOPSIS"
\fBsubtler\fR [GENERIC|MODIFIER] [GROUP] [ACTION]
.
.SH "DESCRIPTION"
\fBsubtler\fR is a commandline tool that allows to interact with a \fBEWMH/NetWM\fR compatible window manager and especially with subtle\. It uses many features of subtle like adding new tags \- but other stuff will also work with every other compliant wm\.
.
.SH "FORMATS"
.
.IP "\(bu" 4
\fBDISPLAY\fR :\fIdisplay number\fR
.
.IP "\(bu" 4
\fBID\fR \fInumber\fR
.
.IP "\(bu" 4
\fBGEOMETRY\fR <xx\fIy\fR+\fIwidth\fR+\fIheight\fR
.
.IP "\(bu" 4
\fBPATTERN\fR
.
.br
Matching works either via plain, regex(7) or window id if applicable\. If a pattern matches more than once ONLY the first match will be used\.
.
.IP
If the \fBPATTERN\fR is \- subtler will read from stdin\.
.
.IP "" 0
.
.SH "OPTIONS"
.
.SH "GENERIC"
.
.IP "\(bu" 4
\fBd\fR, \fB\-\-display\fR DISPLAY
.
.br
Connect to DISPLAY (default: :0)
.
.IP "\(bu" 4
\fBh\fR, \fB\-\-help\fR
.
.br
Show this help and exit
.
.IP "\(bu" 4
\fBV\fR, \fB\-\-version\fR
.
.br
Show version info and exit
.
.IP "" 0
.
.SH "MODIFIER"
.
.IP "\(bu" 4
\fBr\fR, \fB\-\-reload\fR
.
.br
Reload subtle
.
.IP "\(bu" 4
\fBR\fR, \fB\-\-restart\fR
.
.br
Restart subtle
.
.IP "\(bu" 4
\fBQ\fR, \fB\-\-quit\fR
.
.br
Quit subtle
.
.IP "\(bu" 4
\fBC\fR, \fB\-\-current\fR
.
.br
Select current active window/view
.
.IP "\(bu" 4
\fBX\fR, \fB\-\-select\fR
.
.br
Select a window via pointer
.
.IP "" 0
.
.SH "GROUPS"
.
.IP "\(bu" 4
\fBc\fR, \fB\-\-client\fR
.
.br
Use client group
.
.IP "\(bu" 4
\fBg\fR, \fB\-\-gravity\fR
.
.br
Use gravity group
.
.IP "\(bu" 4
\fBe\fR, \fB\-\-screen\fR
.
.br
Use screen group
.
.IP "\(bu" 4
\fBs\fR, \fB\-\-sublet\fR
.
.br
Use sublet group
.
.IP "\(bu" 4
\fBt\fR, \fB\-\-tag\fR
.
.br
Use tag group
.
.IP "\(bu" 4
\fBy\fR, \fB\-\-tray\fR
.
.br
Use tray group
.
.IP "\(bu" 4
\fBv\fR, \fB\-\-view\fR
.
.br
Use views group
.
.IP "" 0
.
.SH "ACTIONS FOR CLIENTS"
.
.IP "\(bu" 4
\fBf\fR, \fB\-\-find\fR PATTERN
.
.br
Find client
.
.IP "\(bu" 4
\fBo\fR, \fB\-\-focus\fR PATTERN
.
.br
Set focus to client
.
.IP "\(bu" 4
\fBF\fR, \fB\-\-full\fR
.
.br
Toggle full
.
.IP "\(bu" 4
\fBO\fR, \fB\-\-float\fR
.
.br
Toggle float
.
.IP "\(bu" 4
\fBS\fR, \fB\-\-stick\fR
.
.br
Toggle stick
.
.IP "\(bu" 4
\fBN\fR, \fB\-\-urgent\fR
.
.br
Toggle urgent
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR
.
.br
List all clients
.
.IP "\(bu" 4
\fBT\fR, \fB\-\-tag\fR PATTERN
.
.br
Add tag to client
.
.IP "\(bu" 4
\fBU\fR, \fB\-\-untag\fR PATTERN
.
.br
Remove tag from client
.
.IP "\(bu" 4
\fBG\fR, \fB\-\-tags\fR
.
.br
Show client tags
.
.IP "\(bu" 4
\fBY\fR, \fB\-\-gravity\fR
.
.br
Set client gravity
.
.IP "\(bu" 4
\fBE\fR, \fB\-\-raise\fR
.
.br
Raise client window
.
.IP "\(bu" 4
\fBL\fR, \fB\-\-lower\fR
.
.br
Lower client window
.
.IP "\(bu" 4
\fBk\fR, \fB\-\-kill\fR PATTERN
.
.br
Kill client
.
.IP "" 0
.
.SH "ACTIONS FOR GRAVITIES"
.
.IP "\(bu" 4
\fBa\fR, \fB\-\-add\fR NAME
.
.br
Create new gravity
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR
.
.br
List all gravities
.
.IP "\(bu" 4
\fBf\fR, \fB\-\-find\fR PATTERN
.
.br
Find a gravity
.
.IP "\(bu" 4
\fBk\fR, \fB\-\-kill\fR PATTERN
.
.br
Kill gravity mode
.
.IP "" 0
.
.SH "ACTIONS FOR SCREENS"
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR
.
.br
List all screens
.
.IP "\(bu" 4
\fBf\fR, \fB\-\-find\fR ID
.
.br
Find a screen
.
.IP "" 0
.
.SH "ACTIONS FOR SUBLETS"
.
.IP "\(bu" 4
\fBa\fR, \fB\-\-add\fR FILE
.
.br
Create new sublet
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR
.
.br
List all sublets
.
.IP "\(bu" 4
\fBu\fR, \fB\-\-update\fR
.
.br
Updates value of sublet
.
.IP "\(bu" 4
\fBA\fR, \fB\-\-data\fR
.
.br
Set data of sublet
.
.IP "\(bu" 4
\fBk\fR, \fB\-\-kill\fR PATTERN
.
.br
Kill sublet
.
.IP "" 0
.
.SH "ACTIONS FOR TAGS"
.
.IP "\(bu" 4
\fBa\fR, \fB\-\-add\fR NAME
.
.br
Create new tag
.
.IP "\(bu" 4
\fBf\fR, \fB\-\-find\fR
.
.br
Find all clients/views by tag
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR
.
.br
List all tags
.
.IP "\(bu" 4
\fBI\fR, \fB\-\-clients\fR
.
.br
Show clients with tag
.
.IP "\(bu" 4
\fBk\fR, \fB\-\-kill\fR PATTERN
.
.br
Kill tag
.
.IP "" 0
.
.SH "ACTIONS FOR VIEWS"
.
.IP "\(bu" 4
\fBa\fR, \fB\-\-add\fR NAME
.
.br
Create new view
.
.IP "\(bu" 4
\fBf\fR, \fB\-\-find\fR PATTERN
.
.br
Find a view
.
.IP "\(bu" 4
\fBl\fR, \fB\-\-list\fR List all views
.
.IP "\(bu" 4
\fBT\fR, \fB\-\-tag\fR PATTERN
.
.br
Add tag to view
.
.IP "\(bu" 4
\fBU\fR, `\-\-untag PATTERN
.
.br
Remove tag from view
.
.IP "\(bu" 4
\fBG\fR, \fB\-\-tags\fR Show view tags
.
.IP "\(bu" 4
\fBI\fR, \fB\-\-clients\fR Show clients on view
.
.IP "\(bu" 4
\fBk\fR, \fB\-\-kill\fR VIEW
.
.br
Kill view
.
.IP "" 0
.
.SH "LISTINGS"
Client listing: \fIwindow id\fR [\-\fI] \fIview id\fR \fIgeometry\fR \fIgravity\fR \fIflags\fR \fIname\fR (\fIclass\fR)
.
.br
Gravity listing: \fIgravity id\fR \fIgeometry\fR
.
.br
Screen listing: \fIscreen id\fR \fIgeometry\fR
.
.br
Tag listing: \fIname\fR
.
.br
View listing: \fIwindow id\fR [\-\fR] \fIview id\fR \fIname\fR
.
.SH "EXAMPLES"
subtler \-c \-l List all clients
.
.br
subtler \-t \-a subtle Add new tag \'subtle\'
.
.br
subtler \-v subtle \-T rocks Tag view \'subtle\' with tag \'rocks\'
.
.br
subtler \-c xterm \-G Show tags of client \'xterm\'
.
.br
subtler \-c \-X \-f Select client and show info
.
.br
subtler \-c \-C \-Y 5 Set gravity 5 to current active client
.
.br
subtler \-t \-f term Show every client/view tagged with \'term\'
.
.SH "BUGS"
Report bugs at http://subforge\.org/projects/subtle/issues
.
.br
Homepage: http://subtle\.subforge\.org
.
.SH "COPYRIGHT"
Copyright (c) Christoph Kappel \fIunexist@dorfelite\.net\fR
.
.SH "SEE ALSO"
subtle(1), subtlext(1), sur(1), surserver(1)


NEWS
News
----
The latest news can be found here:

http://unexist.scrapping.cc/projects/subtle/news


contrib/subtler/subtler.rb
#!/usr/bin/ruby
#
# @package subtler
#
# @file Subtle remote
# @copyright (c) 2005-2010 Christoph Kappel <unexist@dorfelite.net>
# @version $Id$
#
# This program can be distributed under the terms of the GNU GPL.
# See the file COPYING.
#

require "getoptlong"
require "subtle/subtlext"

module Subtle # {{{
  class Subtler # {{{

  # Signals
  trap "INT" do
    exit
  end

    ## initialize {{{
    # Intialize class
    ##

    def initialize
      @mod    = nil
      @group  = nil
      @action = nil

      # Getopt options
      @opts = GetoptLong.new(
        # Groups
        [ "--Client",         "-c", GetoptLong::NO_ARGUMENT ],
        [ "--Gravity",        "-g", GetoptLong::NO_ARGUMENT ],
        [ "--Screen",         "-e", GetoptLong::NO_ARGUMENT ],
        [ "--Sublet",         "-s", GetoptLong::NO_ARGUMENT ],
        [ "--Tag",            "-t", GetoptLong::NO_ARGUMENT ],
        [ "--Tray",           "-y", GetoptLong::NO_ARGUMENT ],
        [ "--View",           "-v", GetoptLong::NO_ARGUMENT ],

        # Actions
        [ "--add",            "-a", GetoptLong::NO_ARGUMENT ],
        [ "--kill",           "-k", GetoptLong::NO_ARGUMENT ],
        [ "--find",           "-f", GetoptLong::NO_ARGUMENT ],
        [ "--focus",          "-o", GetoptLong::NO_ARGUMENT ],
        [ "--full",           "-F", GetoptLong::NO_ARGUMENT ],
        [ "--float",          "-O", GetoptLong::NO_ARGUMENT ],
        [ "--stick",          "-S", GetoptLong::NO_ARGUMENT ],
        [ "--jump",           "-j", GetoptLong::NO_ARGUMENT ],
        [ "--list",           "-l", GetoptLong::NO_ARGUMENT ],
        [ "--tag",            "-T", GetoptLong::NO_ARGUMENT ],
        [ "--untag",          "-U", GetoptLong::NO_ARGUMENT ],
        [ "--tags",           "-G", GetoptLong::NO_ARGUMENT ],
        [ "--retag",          "-A", GetoptLong::NO_ARGUMENT ],
        [ "--clients",        "-I", GetoptLong::NO_ARGUMENT ],
        [ "--views",          "-W", GetoptLong::NO_ARGUMENT ],
        [ "--update",         "-u", GetoptLong::NO_ARGUMENT ],
        [ "--data",           "-D", GetoptLong::NO_ARGUMENT ],
        [ "--gravity",        "-Y", GetoptLong::NO_ARGUMENT ],
        [ "--screen",         "-n", GetoptLong::NO_ARGUMENT ],
        [ "--raise",          "-E", GetoptLong::NO_ARGUMENT ],
        [ "--lower",          "-L", GetoptLong::NO_ARGUMENT ],

        # Modifiers
        [ "--reload_config",  "-r", GetoptLong::NO_ARGUMENT ],
        [ "--reload_sublets", "-R", GetoptLong::NO_ARGUMENT ],
        [ "--restart",        "-q", GetoptLong::NO_ARGUMENT ],
        [ "--quit",           "-Q", GetoptLong::NO_ARGUMENT ],
        [ "--current",        "-C", GetoptLong::NO_ARGUMENT ],
        [ "--select",         "-X", GetoptLong::NO_ARGUMENT ],

        # Other
        [ "--help",           "-h", GetoptLong::NO_ARGUMENT ],
        [ "--version",        "-V", GetoptLong::NO_ARGUMENT ]
      )
    end # }}}

    ## run {{{
    # Run subtler
    ##

    def run
      # Parse arguments
      @opts.each do |opt, arg|
        case opt
          # Groups
          when "--Client"         then @group  = Subtlext::Client
          when "--Gravity"        then @group  = Subtlext::Gravity
          when "--Screen"         then @group  = Subtlext::Screen
          when "--Sublet"         then @group  = Subtlext::Sublet
          when "--Tag"            then @group  = Subtlext::Tag
          when "--Tray"           then @group  = Subtlext::Tray
          when "--View"           then @group  = Subtlext::View

          # Actions
          when "--add"            then @action = :new
          when "--kill"           then @action = :kill
          when "--find"           then @action = :find
          when "--focus"          then @action = :focus
          when "--full"           then @action = :toggle_full
          when "--float"          then @action = :toggle_float
          when "--stick"          then @action = :toggle_stick
          when "--jump"           then @action = :jump
          when "--list"           then @action = :all
          when "--tag"            then @action = :tag
          when "--untag"          then @action = :untag
          when "--tags"           then @action = :tags
          when "--retag"          then @action = :retag
          when "--clients"        then @action = :clients
          when "--views"          then @action = :views
          when "--update"         then @action = :update
          when "--data"           then @action = :data=
          when "--gravity"        then @action = :gravity=
          when "--screen"         then @action = :screen=
          when "--raise"          then @action = :raise
          when "--lower"          then @action = :lower

          # Modifiers
          when "--reload_config"  then @mod = :reload_config
          when "--reload_sublets" then @mod = :reload_sublets
          when "--restart"        then @mod = :restart
          when "--quit"           then @mod = :quit
          when "--current"        then @mod = :current
          when "--select"         then @mod = :select

          # Other
          when "--help"
            usage(@group)
            exit
          when "--version"
            version
            exit
          else
            usage
            exit
        end
      end

      # Get arguments
      arg1 = ARGV.shift
      arg2 = ARGV.shift

      # Pipes?
      arg1 = ARGF.read.chop if("-" == arg1)
      arg2 = ARGF.read.chop if("-" == arg2)

      # Modifiers
      case @mod
        when :reload_config  then  Subtlext::Subtle.reload_config
        when :reload_sublets then  Subtlext::Subtle.reload_sublets
        when :restart        then  Subtlext::Subtle.restart
        when :quit           then  Subtlext::Subtle.quit
        when :current
          arg2 = arg1
          arg1 = :current
        when :select
          arg2 = arg1
          arg1 = Subtlext::Subtle.select_window
      end

      # Call method
      unless(@group.nil? || @action.nil?)
        # Check singleton and instance methods
        if((@group.singleton_methods << :new).include?(@action))
          obj   = @group
          arg   = arg1
          arity = obj.method(@action).arity
        elsif(@group.instance_methods.include?(@action))
          obj   = @group.send(:find, arg1)
          arg   = arg2
          arity = @group.instance_method(@action).arity
        end

        # Check arity
        case arity
          when 1
            ret = obj.send(@action, arg)
          when -1
            if([ Subtlext::View, Subtlext::Tag].include?(@group))
              # Create new object
              ret = obj.send(@action, arg)
              ret.save
            end

            exit
          when nil
            usage(@group)
            exit
          else
            ret = obj.send(@action)
        end

        # Handle result
        case ret
          when Array
            ret.each do |r|
              printer(r)
            end
          else
            printer(ret)
        end
      else
        usage(@group)
      end
    end # }}}

    private

    # Client modes
    MODE_FULL  = (1 << 1)
    MODE_FLOAT = (1 << 2)
    MODE_STICK = (1 << 3)

    def printer(value) # {{{
      case value
        when Subtlext::Client # {{{
          puts "%#10x %s %d %4d x %4d + %4d + %-4d %12.12s %d %s%s%s %s (%s)" % [
            value.win,
            value.views.map { |v| v.name }.include?(Subtlext::View.current.name) ? "*" : "-",
            Subtlext::View.current.id,
            value.geometry.x,
            value.geometry.y,
            value.geometry.width,
            value.geometry.height,
            value.gravity.name,
            value.screen.id,
            (value.flags & MODE_FULL).zero?  ? "-" : "F",
            (value.flags & MODE_FLOAT).zero? ? "-" : "O",
            (value.flags & MODE_STICK).zero? ? "-" : "S",
            value.instance,
            value.klass
          ] # }}}
        when Subtlext::Gravity # {{{
          puts "%15s %3d x %-3d %3d + %-3d" % [
            value.name,
            value.geometry.x,
            value.geometry.y,
            value.geometry.width,
            value.geometry.height,
          ] # }}}
        when Subtlext::Screen # {{{
          puts "%d %4d x %-4d %4d + %-4d" % [
            value.id,
            value.geometry.x,
            value.geometry.y,
            value.geometry.width,
            value.geometry.height,
          ] # }}}
        when Subtlext::Tag # {{{
          puts value.name # }}}
        when Subtlext::Tray # {{{
          puts "%#10lx %s (%s)" % [
            value.win,
            value.instance,
            value.klass
          ] # }}}
        when Subtlext::View # {{{
          puts "%#10x %s %2d %s" % [
            value.win,
            value.current? ? "*" : "-",
            value.id,
            value.name
          ] # }}}
      end
    end # }}}

    def usage(group) # {{{
      puts "Usage: subtler [GENERIC|MODIFIER] [GROUP] [ACTION]\n\n"

      if(group.nil?)
        puts <<EOF

Generic:
  -d, --display=DISPLAY   Connect to DISPLAY (default: #{ENV["DISPLAY"]})
  -D, --debug             Print debugging messages
  -h, --help              Show this help and exit
  -V, --version           Show version info and exit

Modifier:
  -r, --reload-config     Reload config
  -R, --reload-sublets    Reload sublets
  -q, --restart           Restart subtle
  -Q, --quit              Quit %s
  -C, --current           Select current active window/view
  -X, --select            Select a window via pointer

Groups:
  -c, --client            Use client group
  -g, --gravity           Use gravity group
  -e, --screen            Use screen group
  -s, --sublet            Use sublet group
  -t, --tag               Use tag group
  -y, --tray              Use tray group
  -v, --view              Use views group
EOF
      end

      if(group.nil? or Subtlext::Client == group)
        puts <<EOF

Actions for clients:
  -f, --find=PATTERN      Find client
  -o, --focus=PATTERN     Set focus to client
  -F, --full              Toggle full
  -O, --float             Toggle float
  -S, --stick             Toggle stick
  -l, --list              List all clients
  -T, --tag=PATTERN       Add tag to client
  -U, --untag=PATTERN     Remove tag from client
  -G, --tags              Show client tags
  -Y, --gravity           Set client gravity
  -n, --screen            Set client screen
  -E, --raise             Raise client window
  -L, --lower             Lower client window
  -k, --kill=PATTERN      Kill client
EOF
      end

      if(group.nil? or Subtlext::Gravity == group)
        puts <<EOF

Actions for gravities:
  -a, --add=NAME          Create new gravity
  -l, --list              List all gravities
  -f, --find=PATTERN      Find a gravity
  -k, --kill=PATTERN      Kill gravity mode
EOF
      end

      if(group.nil? or Subtlext::Screen == group)
        puts <<EOF

Actions for screens:
  -l, --list              List all screens
  -f, --find=ID           Find a screen
EOF
      end

      if(group.nil? or Subtlext::Sublet == group)
        puts <<EOF

Actions for sublets:
  -a, --add=FILE          Create new sublet
  -l, --list              List all sublets
  -u, --update            Updates value of sublet
  -A, --data              Set data of sublet
  -k, --kill=PATTERN      Kill sublet
EOF
      end

      if(group.nil? or Subtlext::Tag == group)
        puts <<EOF

Actions for tags:
  -a, --add=NAME          Create new tag
  -f, --find              Find all clients/views by tag
  -l, --list              List all tags
  -I, --clients           Show clients with tag
  -k, --kill=PATTERN      Kill tag
EOF
      end

      if(group.nil? or Subtlext::View == group)
        puts <<EOF

Actions for views:
  -a, --add=NAME          Create new view
  -f, --find=PATTERN      Find a view
  -l, --list              List all views
  -T, --tag=PATTERN       Add tag to view
  -U, --untag=PATTERN     Remove tag from view
  -G, --tags              Show view tags
  -I, --clients           Show clients on view
  -k, --kill=VIEW         Kill view
EOF
      end

      puts <<EOF

Formats:
  DISPLAY:  :<display number>
  ID:       <number>
  GEOMETRY: <x>x<y>+<width>+<height>
  PATTERN:
    Matching works either via plaintext, regex (see regex(7)), id or window id
    if applicable. If a pattern matches more than once ONLY the first match will
    be used.
    If the PATTERN is '-' subtler will read from stdin.

Listings:
  Client listing:  <window id> [-*] <view id> <geometry> <gravity> <screen> <flags> <name> (<class>)
  Gravity listing: <gravity id> <geometry>
  Screen listing:  <screen id> <geometry>
  Tag listing:     <name>
  View listing:    <window id> [-*] <view id> <name>

Examples:
  subtler -c -l                List all clients
  subtler -t -a subtle         Add new tag 'subtle'
  subtler -v subtle -T rocks   Tag view 'subtle' with tag 'rocks'
  subtler -c xterm -G          Show tags of client 'xterm'
  subtler -c -x -f             Select client and show info
  subtler -c -C -y 5           Set gravity 5 to current active client
  subtler -t -f term           Show every client/view tagged with 'term'

Please report bugs to <unexist@dorfelite.net>
EOF
    end # }}}

    def version # {{{
      puts <<EOF
subtler #{Subtlext::VERSION} - Copyright (c) 2005-2010 Christoph Kappel
Released under the GNU General Public License
Using Ruby #{RUBY_VERSION}
EOF
    end # }}}
  end # ]]]
end # }}}

# vim:ts=2:bs=2:sw=2:et:fdm=marker
